packages/fuels-core/src/code_gen/flat_abigen.rs

```diff
@@ -0,0 +1,261 @@
+                            self
+                        }
+
+                        pub fn wallet(&mut self, wallet: LocalWallet) -> &mut Self {
+                            self.wallet = wallet;
+                            self
+                        }
+
+                        pub fn build(self) -> #name {
+                            #name { contract_id: self.contract_id, wallet: self.wallet }
+                        }
+                    }
+                },
+            )
+        };
+
+        Ok(quote! {
+            pub use #name_mod::*;
+
+            #[allow(clippy::too_many_arguments)]
+            pub mod #name_mod {
+                #![allow(clippy::enum_variant_names)]
+                #![allow(dead_code)]
+                #![allow(unused_imports)]
+
+                #includes
+
+                #code
+
+                #abi_structs
+                #abi_enums
+            }
+        })
+    }
+
+    pub fn functions(&self) -> Result<TokenStream, Error> {
+        let mut tokenized_functions = Vec::new();
+
+        for function in &self.abi.functions {
+            let tokenized_fn = _new_expand_function(function, &self.types)?;
+            tokenized_functions.push(tokenized_fn);
+        }
+
+        Ok(quote! { #( #tokenized_functions )* })
+    }
+
+    fn abi_structs(&self) -> Result<TokenStream, Error> {
+        let mut structs = TokenStream::new();
+
+        // Prevent expanding the same struct more than once
+        let mut seen_struct: Vec<&str> = vec![];
+
+        for prop in &self.abi.types {
+            // If it isn't a struct, skip.
+            if !prop.is_struct_type() {
+                continue;
+            }
+
+            // Skip custom type generation if the custom type is a Sway-native type.
+            // This means ABI methods receiving or returning a Sway-native type
+            // can receive or return that native type directly.
+            if FlatAbigen::is_sway_native_type(&prop.type_field) {
+                continue;
+            }
+
+            if !seen_struct.contains(&prop.type_field.as_str()) {
+                structs.extend(_new_expand_custom_struct(prop, &self.types)?);
+                seen_struct.push(&prop.type_field);
+            }
+        }
+
+        Ok(structs)
+    }
+
+    // Checks whether the given type field is a Sway-native type.
+    // It's expected to come in as `"struct T"` or `"enum T"`.
+    // `T` is a Sway-native type if it matches exactly one of
+    // the reserved strings, such as "Address" or "ContractId".
+    fn is_sway_native_type(type_field: &str) -> bool {
+        let split: Vec<&str> = type_field.split_whitespace().collect();
+
+        if split.len() > 2 {
+            return false;
+        }
+        split[1] == CONTRACT_ID_SWAY_NATIVE_TYPE || split[1] == ADDRESS_SWAY_NATIVE_TYPE
+    }
+
+    fn abi_enums(&self) -> Result<TokenStream, Error> {
```

Comment:
Why don't we have the same `seen` mechanism as we do for structs?

Reply:
Unrelated to the changes introduced here (this was mostly copy-pasted from `abigen.rs`), but it should have the same duplicate detection logic, and it might have slipped through reviews. Easy fix though.

Reply:
Opened #542